ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -790,25 +797,62 @@ let choose_host_uses_wlb ~__context =
       (Db.Pool.get_other_config ~__context
          ~self:(Helpers.get_pool ~__context)))
 
+
 (** Given a virtual machine, returns a host it can boot on, giving   *)
 (** priority to an affinity host if one is present. WARNING: called  *)
 (** while holding the global lock from the message forwarding layer. *)
 let choose_host_for_vm ~__context ~vm ~snapshot =
   if choose_host_uses_wlb ~__context then
     try
+      let current_time = Unix.time() in
+      let filter_calm_down value =
+        List.filter (fun e -> current_time -. e <= !Xapi_globs.wlb_host_cool_down_time) value
+      in
+      let rec pow base = function
+        | 0 -> 1.0
+        | 1 -> base
+        | n -> base *. (pow base (n-1))
+      in
       let possible_rec (host, recommendation) =
         match recommendation with
         | Recommendation {source; id; score} ->
             Some (host, score, id)
         | _ ->
             None
       in
+      let cool_down_host (host, score, id) =
+        match Cool_down_map.find_opt host !cool_down_data with
+        | None -> (host, score, id)
+        | Some times ->
+          let penalty = List.length times |> pow !Xapi_globs.wlb_host_cool_down_penalty in
+          let penaltied_value = score *. penalty in
+          (host, (penaltied_value), id)
+      in
+      let heat_host = function
+        | Some times -> Some (current_time::times)
```

Comment:
Can this list become very long? Should it have an upper-bound?

Reply:
The list contains the VM starts on the host during last N second, (N configurable with default value 10, it will not be huge)